zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+
+        let qrange = self.queued.range(current_range).rev();
+        // We could try to calculate the correct range here, but it's easier to
+        // just limit the heights using `qrange.zip()`. We also make sure
+        // BlockHeight can't underflow.
+        let expected_heights = successors(Some(next_checkpoint_height), |n| {
+            n.0.checked_sub(1).map(BlockHeight)
+        });
+
+        // The hash we expect this block to have
+        let mut expected_hash = next_checkpoint_hash;
+        // The height of the last block we checked
+        let mut last_height = None;
+
+        for ((&height, qblocks), expected_height) in qrange.zip(expected_heights) {
+            // Check that the heights are continuous
+            if height != expected_height {
+                // There is a gap in the block chain
+                // Missing blocks are not an error - wait for more blocks
+                return (Some(expected_height), Some(expected_hash));
+            }
+
+            // Check if any queued block at this height is part of the chain
+            // TODO(teor): consider using find_map() with a qblock method
+            let parent_hash = qblocks
+                .iter()
+                .skip_while(|qblock| qblock.hash != expected_hash)
+                .map(|qblock| qblock.block.header.previous_block_hash)
+                .next();
+
+            if let Some(parent_hash) = parent_hash {
+                expected_hash = parent_hash;
+                last_height = Some(height);
+            } else {
+                // There was no valid block in the queued block list
+                // Missing blocks are not an error - wait for more blocks
+                //
+                // This check also handles zero-length vecs, which should be
+                // impossible
+                return (Some(expected_height), Some(expected_hash));
+            }
+        }
+
+        // If the loop didn't execute, we're waiting for the next checkpoint block
+        // TODO(teor): add tests for this case
+        if last_height.is_none() {
+            return (Some(next_checkpoint_height), Some(next_checkpoint_hash));
+        }
+
+        // At this point, last_height is the height of the last block, but
+        // expected_hash is the hash of the last block's parent block.
+        //
+
+        // If we're at the genesis block, we want to return None for the parent
+        // block height.
+        let parent_height = last_height.unwrap().0.checked_sub(1).map(BlockHeight);
+        (parent_height, Some(expected_hash))
+    }
+
+    /// Check all the blocks in the current checkpoint range. Send `Ok` for the
+    /// blocks that are in the chain, and `Err` for side-chain blocks.
+    ///
+    /// Does nothing if verification has finished.
+    ///
+    /// Check that the current checkpoint range is complete before calling this
+    /// function.
+    fn submit_current_checkpoint_range(&mut self) {
```

Comment:
could we call the check as part of this fn rather than making calling it separately an invariant the user has to uphold?

Reply:
Fixed in 75c550613a5aa172dc0c4cad5ebf58f175766c8f